zone.go

```diff
@@ -0,0 +1,244 @@
+package dns
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"sort"
+	"strings"
+)
+
+// AddDomain adds origin to s if s is not already a FQDN.
+func AddDomain(origin, s string) string {
+	// If you want the result to end in a ".", pass an origin that ends in a "."
+	// ("origin.", "foo") -> "foo.origin."
+	// ("origin", "foo") -> "foo.origin"
+	// ("origin.", "foo.") -> "foo." (already a FQDN)
+	// ("origin.", "@") -> "origin." (@ represents the apex domain)
+	// ("origin.", "") -> "origin." (not obvious)
+	if len(origin) == 0 || origin[0] == '.' {
+		log.Fatalf("Invalid origin: %#v\n", origin)
+	}
+	if s == "@" || len(s) == 0 {
+		return origin
+	}
+	if s == "*" {
+		return "*." + origin
+	}
+	if s[len(s)-1] == '.' {
+		return s
+	}
+	return s + "." + origin
+}
+
+// TrimDomain trims the origin from a FQDN if it exists.
+func TrimDomain(origin, s string) string {
+	// If the return value ends in a ".", the domain was not the suffix.
+
+	if len(s) == 0 || s == "." {
+		return "."
+	}
+	name := s[:]
+	if name[len(name)-1] == '.' {
+		name = name[:len(name)-1]
+	}
+	for name[0] == '.' {
+		name = name[1:]
+	}
+
+	suffix := origin[:]
+	if origin[len(origin)-1] == '.' {
+		suffix = origin[:len(origin)-1]
+	}
+	if name == suffix {
+		return "@"
+	}
+	return strings.TrimSuffix(name, "."+suffix)
+}
+
+type Zone struct {
```

Comment:
It only needs a struct so that it can be sorted. I could create a simple `type sortableRR []RR` type just to create an interface that matches Sort.  I could re-factor the code to not need Origin and DefaultTtl by assuming the RRs have FQDNs and proper TTL set already.